borg/platform.py

```diff
@@ -14,3 +18,86 @@ def acl_get(path, item, st, numeric_owner=False):
 
     def acl_set(path, item, numeric_owner=False):
         pass
+
+# GenericDirEntry, scandir_generic (c) 2012 Ben Hoyt
+# from the python-scandir package (3-clause BSD license, just like us, so no troubles here)
+# note: simplified version
+
+
+class GenericDirEntry(object):
+    __slots__ = ('name', '_stat', '_lstat', '_scandir_path', '_path')
+
+    def __init__(self, scandir_path, name):
+        self._scandir_path = scandir_path
+        self.name = name
+        self._stat = None
+        self._lstat = None
+        self._path = None
+
+    @property
+    def path(self):
+        if self._path is None:
+            self._path = os.path.join(self._scandir_path, self.name)
+        return self._path
+
+    def stat(self, follow_symlinks=True):
+        assert not follow_symlinks
+        return os.lstat(self.path)
+
+    def is_dir(self, follow_symlinks=True):
+        assert not follow_symlinks
+        try:
+            st = self.stat(False)
+        except OSError as e:
+            if e.errno != errno.ENOENT:
+                raise
+            return False  # Path doesn't exist or is a broken symlink#
+        return stat.S_ISDIR(st.st_mode)
+
+    def is_file(self, follow_symlinks=True):
+        assert not follow_symlinks
+        try:
+            st = self.stat(False)
+        except OSError as e:
+            if e.errno != errno.ENOENT:
+                raise
+            return False  # Path doesn't exist or is a broken symlink
+        return stat.S_ISREG(st.st_mode)
+
+    def is_symlink(self):
+        try:
+            st = self.stat(False)
+        except OSError as e:
+            if e.errno != errno.ENOENT:
+                raise
+            return False  # Path doesn't exist or is a broken symlink
+        return stat.S_ISLNK(st.st_mode)
+
+    def inode(self):
+        st = self.stat(False)
+        return st.st_ino
+
+    def __str__(self):
+        return '<{0}: {1!r}>'.format(self.__class__.__name__, self.name)
+
+    __repr__ = __str__
+
+
+def scandir_generic(path='.'):
+    """Like os.listdir(), but yield DirEntry objects instead of returning a list of names."""
+    for name in sorted(os.listdir(path)):
+        yield GenericDirEntry(path, name)
+
+try:
+    from os import scandir
+except ImportError:
+    try:
+        # Try python-scandir on Python 3.4
+        from scandir import scandir
+    except ImportError:
+        # If python-scandir is not installed, then use a version that is just as slow as listdir.
+        scandir = scandir_generic
+
+
+def scandir_inorder(path='.'):
+    return sorted(scandir(path), key=lambda dirent: dirent.inode())
```

Comment:
there is also helpers. ;)
which we will tear into pieces when the time has come.